submarine/src/main/java/org/apache/zeppelin/submarine/SubmarineInterpreter.java

```diff
@@ -0,0 +1,275 @@
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.CHECKPOINT_PATH;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.CLEAN_CHECKPOINT;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.COMMAND_CLEAN;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.COMMAND_JOB_RUN;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.COMMAND_JOB_SHOW;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.COMMAND_TYPE;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.COMMAND_USAGE;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.INPUT_PATH;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.MACHINELEARING_DISTRIBUTED_ENABLE;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.OPERATION_TYPE;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.PS_LAUNCH_CMD;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.SUBMARINE_ALGORITHM_HDFS_PATH;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.TF_CHECKPOINT_PATH;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.USERNAME_SYMBOL;
+import static org.apache.zeppelin.submarine.commons.SubmarineConstants.WORKER_LAUNCH_CMD;
+import static org.apache.zeppelin.submarine.commons.SubmarineUtils.unifyKey;
+
+/**
+ * SubmarineInterpreter of Hadoop Submarine implementation.
+ * Support for Hadoop Submarine cli. All the commands documented here
+ * https://github.com/apache/hadoop/tree/trunk/hadoop-submarine/hadoop-submarine-core
+ * /src/site/markdown/QuickStart.md is supported.
+ */
+public class SubmarineInterpreter extends Interpreter {
+  private Logger LOGGER = LoggerFactory.getLogger(SubmarineInterpreter.class);
+
+  // Number of submarines executed in parallel for each interpreter instance
+  protected int concurrentExecutedMax = 1;
+
+  private boolean needUpdateConfig = true;
+  private String currentReplName = "";
+
+  SubmarineContext submarineContext = null;
+
+  public SubmarineInterpreter(Properties properties) {
+    super(properties);
+
+    String concurrentMax = getProperty(SubmarineConstants.SUBMARINE_CONCURRENT_MAX, "1");
+    concurrentExecutedMax = Integer.parseInt(concurrentMax);
+
+    submarineContext = SubmarineContext.getInstance();
+  }
+
+  @Override
+  public void open() {
+    LOGGER.info("SubmarineInterpreter open()");
+  }
+
+  @Override
+  public void close() {
+    submarineContext.stopAllSubmarineJob();
+  }
+
+  private void setParagraphConfig(InterpreterContext context) {
+    String replName = context.getReplName();
+    if (StringUtils.equals(currentReplName, replName)) {
+      currentReplName = context.getReplName();
+      needUpdateConfig = true;
+    }
+    if (needUpdateConfig) {
+      needUpdateConfig = false;
+      if (currentReplName.equals("submarine") || currentReplName.isEmpty()) {
```

Comment:
IIRC, the interpreter name is configurable? eg.  spark is `spark2` in some config. so likely you shouldn't hard code `submarine`

Reply:
I am at https://github.com/apache/zeppelin/pull/3271 , Has all interpreters been able to customize some configurations.
Add this code to the submarine interpreter for a better experience.
Will not have any effect on other interpreters.

Reply:
I mean like install any interpreter with a different name `--name` in
https://zeppelin.apache.org/docs/0.8.0/usage/interpreter/installation.html#spark-interpreter-for-scala-210-and-211

or just name it something different - the "Name" box can say "sub" and the "Interpreter" box can be "submarine"
https://zeppelin.apache.org/docs/0.8.0/usage/interpreter/overview.html#what-is-zeppelin-interpreter

Reply:
Ok, I am going to change the way to determine if it is a submarine interpreter.